scope.go

```diff
@@ -269,16 +270,16 @@ func (s *scope) reportRegistryWithLock() {
 		}
 		s.cachedReporter.Flush()
 	}
-	s.registry.RUnlock()
 }
 
 func (s *scope) Counter(name string) Counter {
 	name = s.sanitizer.Name(name)
-	s.cm.RLock()
+
+	s.cm.Lock()
+	defer s.cm.Unlock()
```

Comment:
This could have large performance implications, since an `RLock` allows multiple readers, while `Lock` only allows one reader.

I think it's common for callers to not hold a reference to a `Counter`, and to rely on this cache, and changing to a lock can significantly increase contention.

I'd suggest changing this to:
```go
func (s *scope) getCounter(name string) (Counter, bool) {
  s.cm.RLock()
  defer s.cm.RUnlock()

  val, ok := s.counters[name]
  return val, ok
}

func (s *scope) Counter(name string) Counter {
  if c, ok := s.getCounter(name); ok {
    return c
  }

  s.cm.Lock()
  defer s.cm.Unlock()

   if val,  ok := s.counters[name]; ok {
    return val
  }
  
  var cachedCounter CachedCount
  val = newCounter(cachedCounter)
  s.counters[name] = val
  return val
}
```

as a bonus, it significantly reduces the amount of nesting,
https://github.com/uber-go/guide/blob/master/style.md#reduce-nesting

same is true for `Gauge`, `Timer`, etc below

Reply:
I explained in the description that benchmarks are negligibly impacted. Also, pedantically, `getX` is not a recommended Go naming convention:

> it's neither idiomatic nor necessary to put Get into the getter's name

(ref: [Effective Go](https://golang.org/doc/effective_go.html#Getters))

I'll make the change recommended here to minimize internal version churn, but let's put any additional changes into subsequent PRs. Sound fair?

Reply:
Is there benchmarks that test the contention caused by concurrency? I think we'll only see impact from benchmarks in that scenario.

Agree that lack of nesting can be done separately, just want to make sure we keep using `RLock` here, thanks